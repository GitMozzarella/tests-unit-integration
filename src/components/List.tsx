import { useState, useMemo } from 'react'
import { Item } from './Item'
import { Task } from 'src/utils/types'

type Props = {
	items: Task[]
	onDelete: (id: Task['id']) => void
	onToggle: (id: Task['id']) => void
}

export const List = ({ items, onDelete, onToggle }: Props) => {
	const [filter, setFilter] = useState<boolean>(false)

	const filteredItems = useMemo(() => {
		return filter ? items.filter(item => !item.done) : items
	}, [filter, items])

	const toggleFilter = () => setFilter(prevFilter => !prevFilter)

	const getButtonText = () => {
		return filter ? 'Все задачи' : 'Невыполненные задачи'
	}

	const buttonClass = `button ${filter ? 'button-all' : 'button-active'}`

	return (
		<>
			<button onClick={toggleFilter} className={buttonClass}>
				{getButtonText()}
			</button>
			<ul className='task-list tasks' data-testid='task-list'>
				{filteredItems.map(item => (
					<Item
						{...item}
						key={item.id}
						onDelete={onDelete}
						onToggle={onToggle}
					/>
				))}
			</ul>
		</>
	)
}
