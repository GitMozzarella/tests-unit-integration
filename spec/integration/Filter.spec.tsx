import ue from '@testing-library/user-event'
import { App } from '../../src/App'
import { List } from '../../src/components/List'

import { render, screen } from '@testing-library/react'
import { Task } from 'src/utils/types'

const userEvent = ue.setup({
	advanceTimers: jest.advanceTimersByTime
})

describe('Список задач', () => {
	it('С включенным фильтром', async () => {
		const onDelete = jest.fn()
		const onToggle = jest.fn()

		const items: Task[] = [
			{ id: '325235', header: 'убраться дома', done: false },
			{ id: '22377', header: 'приготовить ужин', done: false },
			{ id: '42', header: 'выполнить дз', done: false }
		]

		render(<App />)
		const { rerender, asFragment } = render(
			<List items={items} onDelete={onDelete} onToggle={onToggle} />
		)

		const screenFirstRender = asFragment()

		await userEvent.click(screen.getByText(/Невыполненные задачи/i))

		expect(screen.queryByText('доделать отчет по работе')).toBeNull()

		rerender(<List items={items} onDelete={onDelete} onToggle={onToggle} />)

		const screenSecondRender = asFragment()

		expect(screen.queryByText('доделать отчет по работе')).toBeNull()
		expect(screen.getByText('убраться дома')).toBeInTheDocument()
		expect(screen.getByText('выполнить дз')).toBeInTheDocument()
		expect(screen.getByText('приготовить ужин')).toBeInTheDocument()

		expect(screenFirstRender).toMatchDiffSnapshot(screenSecondRender)
	})

	it('C выключенным фильтром', async () => {
		const onDelete = jest.fn()
		const onToggle = jest.fn()

		const items: Task[] = [
			{ id: '325235', header: 'убраться дома', done: true },
			{ id: '22377', header: 'приготовить ужин', done: false },
			{ id: '42', header: 'выполнить дз', done: true }
		]

		render(<List items={items} onDelete={onDelete} onToggle={onToggle} />)

		await userEvent.click(screen.getByText(/Невыполненные задачи/i))

		expect(screen.queryByText('убраться дома')).toBeNull()
		expect(screen.queryByText('выполнить дз')).toBeNull()

		await userEvent.click(screen.getByText(/Все задачи/i))

		expect(screen.getByText('убраться дома')).toBeInTheDocument()
		expect(screen.getByText('приготовить ужин')).toBeInTheDocument()
		expect(screen.getByText('выполнить дз')).toBeInTheDocument()
	})

	it('Фильтрация с учетом частичного совпадения заголовка', async () => {
		const onDelete = jest.fn()
		const onToggle = jest.fn()

		const items: Task[] = [
			{ id: '1', header: 'покупка молока', done: false },
			{ id: '2', header: 'приготовить ужин', done: false },
			{ id: '3', header: 'выполнить домашку', done: true }
		]

		const { rerender } = render(
			<List items={items} onDelete={onDelete} onToggle={onToggle} />
		)

		expect(screen.getByText('покупка молока')).toBeInTheDocument()
		expect(screen.getByText('приготовить ужин')).toBeInTheDocument()
		expect(screen.getByText('выполнить домашку')).toBeInTheDocument()

		await userEvent.click(screen.getByText(/Невыполненные задачи/i))

		expect(screen.getByText('покупка молока')).toBeInTheDocument()
		expect(screen.getByText('приготовить ужин')).toBeInTheDocument()
		expect(screen.queryByText('выполнить домашку')).toBeNull()

		const updatedItems: Task[] = [
			{ id: '4', header: 'молоко и хлеб', done: false },
			{ id: '5', header: 'уточнить ужин', done: true }
		]

		rerender(
			<List items={updatedItems} onDelete={onDelete} onToggle={onToggle} />
		)

		expect(screen.getByText('молоко и хлеб')).toBeInTheDocument()
		expect(screen.queryByText('уточнить ужин')).toBeNull()
	})
})
