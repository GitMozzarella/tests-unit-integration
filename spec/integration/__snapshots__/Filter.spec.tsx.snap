// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Список задач С включенным фильтром 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,10 +1,10 @@
  <DocumentFragment>
    <button
-     class="button button-active"
+     class="button button-all"
    >
-     Невыполненные задачи
+     Все задачи
    </button>
    <ul
      class="task-list tasks"
      data-testid="task-list"
    >"
`;
